File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/cats/AttemptNonRaised.scala
package arrow.benchmarks.effects.scala.cats

import cats.effect.IO

object AttemptNonRaised {

  def ioLoopHappy(size: Int, i: Int): IO[Int] =
    if (i < size) {
      IO {
        i + 1
      }.attempt
        .flatMap { result =>
          result.fold[IO[Int]](IO.raiseError[Int], ioLoopHappy(size, _))
        }
    } else IO.pure(1)

}

File: modules/benchmarks/arrow-kio-benchmarks/src/main/kotlin/kio/AttemptNonRaised.kt
package arrow.benchmarks.effects.kio

import it.msec.kio.Task
import it.msec.kio.attempt
import it.msec.kio.effect
import it.msec.kio.just
import it.msec.kio.recover
import it.msec.kio.flatMap
import it.msec.kio.result.getOrThrow
import it.msec.kio.runtime.Runtime

object AttemptNonRaised {

  fun attemptNonRaised(size: Int): Int = Runtime.unsafeRunSync(loopHappy(size, 0)).getOrThrow()

  fun loopHappy(size: Int, i: Int): Task<Int> =
    if (i < size) {
      effect { i + 1 }
        .attempt()
        .recover { throw RuntimeException("not happening") }
        .flatMap { loopHappy(size, it) }
    } else
      just(1)
}
